### ISSUE TITLE
Cannot override --flags= on command line

### ISSUE THREAD

**Issue by kolmodin** [haskell/cabal#4452]
**WARNING** There are updates to this ticket, please read the discussion.

**Summary.** It's common that the last given value of a flag "wins".

```
$ cabal configure -fdev -v
...
Flags chosen: dev=True
...

$ cabal configure -f-dev -v
...
Flags chosen: dev=False
....

$ cabal configure -fdev -f-dev -v
...
Flags chosen: dev=True
...
```

I expected the last line to have set `dev=False`. According to the [documentation](https://www.haskell.org/cabal/users-guide/installing-packages.html#controlling-flag-assignments) it should be `dev=False`.

**Explanation (by @ezyang).**

Cabal has quite flexible handling of duplicate instances of arguments when doing command line handling. It does so using *monoids*, which are types which have a binary "combining" operation defined for them. ConfigFlags, defined in `Cabal/Distribution/Simple/Setup.hs`, defines the monoid instance for ALL configuration flags, by applying the monoid for each individual field.

Flag assignment is recorded in `configConfigurationsFlags`, which is defined to be a list. The monoid instance for lists is append. This means that when a flag shows up twice, the lists get appended together; with the example above, you end up with `[("dev", True), ("dev", False)]`.

The bug is in this code `Distribution/PackageDescription/Configuration.hs`:

```
    d2c manual n b = case lookup n userflags of
                     Just val -> [val]
                     Nothing
                      | manual -> [b]
                      | otherwise -> [b, not b]
```

Here, `userflags` is the list of flag assignments, possibly with duplicates, in the order they occurred on the command line. The problem is that `lookup` will pick up the FIRST entry, rather than the LAST entry. A better and more efficient strategy is to build a map, preferring the last entry, and then do a lookup on this map. If you do that, the bug will be fixed!

**What you will learn.** How to build Cabal. How to convert a key-value list of items into a map, and then do a lookup on it. How to do command line arguments with Monoids. How to look through the verbose output of cabal configure to find the info you need.

**Going further.** This bug doesn't really require a test, but if you want to add one, it should be fairly easy. Follow the instructions in https://github.com/haskell/cabal/tree/master/cabal-testsuite

**Comment by kolmodin** [comment]
```
$ cabal --version
cabal-install version 1.25.0.0
compiled using version 1.25.0.0 of the Cabal library 
```

**Comment by 23Skidoo** [comment]
Makes sense.

**Comment by colinwahl** [comment]
Hi everybody.  I have made the change to `Distribution/PackageDescription/Configuration.hs` to use a map instead of a list, but I am having trouble seeing the output 

`$ cabal configure -f-dev -v
...
Flags chosen: dev=False
....`

as shown in the description.  Could anyone point me to where I am going wrong?

**Comment by ezyang** [comment]
Hi @colinwahl! @krpopo has been looking at this ticket and discovered that there are actually two buggy call sites. With the change I suggested, `cabal act-as-setup -- configure -fdev -f-dev -v` works correctly (this call bypasses cabal-install and goes directly to Cabal's entry point), but `cabal configure` does not. In fact, there appears to be *another* `lookup` site in cabal-install (in `cabal-install/Distribution/Solver/Modular/Preference.hs`) which is causing `cabal configure` to continue to be buggy. Fixing that one should fix all the visible bugs, but it's possible we should refactor `FlagAssignment` to be a map rather than a list and solve all occurrences of this bug.

**Comment by krpopo** [comment]
I don't plan to work on this any more

**Comment by 23Skidoo** [comment]
@VetoPlayer is looking into this one.

**Comment by 23Skidoo** [comment]
Fixed by #4940.
